### ISSUE TITLE
Stack smash detected when ring buffer is created

### ISSUE THREAD

**Issue by gafferongames** [libbpf/libbpf#797]
Hey folks, I'm getting a very strange crash in my XDP program and I need some help. 

It's a stack smash inside the libbpf code that attaches the xdp program to the NIC. It seems to be triggered by including the ring buffer create in my program, but the ring buffer create doesn't even run until after the program is attached to the NIC, so ... I'm very confused :)

I reduced my crashing program down to a minimal repro here:

https://github.com/mas-bandwidth/fps/tree/main/004

Just go:

```console
make && sudo ./server
```

and you'll see the crash:

```
glenn@hulk:~/fps/004$ sudo ./server
[sudo] password for glenn:
after xdp_program__open_file
*** stack smashing detected ***: terminated
Aborted
```

If you run it inside lldb, you see:

```console
glenn@hulk:~/fps/004$ sudo lldb ./server
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'lldb.embedded_interpreter'
(lldb) target create "./server"
Current executable set to '/home/glenn/fps/004/server' (x86_64).
(lldb) run
Process 4563 launched: '/home/glenn/fps/004/server' (x86_64)
after xdp_program__open_file
*** stack smashing detected ***: terminated
Process 4563 stopped
* thread #1, name = 'server', stop reason = signal SIGABRT
    frame #0: 0x00007ffff7c969fc libc.so.6`pthread_kill + 300
libc.so.6`pthread_kill:
->  0x7ffff7c969fc <+300>: movl   %eax, %r13d
    0x7ffff7c969ff <+303>: negl   %r13d
    0x7ffff7c96a02 <+306>: cmpl   $0xfffff000, %eax         ; imm = 0xFFFFF000
    0x7ffff7c96a07 <+311>: movl   $0x0, %eax
(lldb) bt
* thread #1, name = 'server', stop reason = signal SIGABRT
  * frame #0: 0x00007ffff7c969fc libc.so.6`pthread_kill + 300
    frame #1: 0x00007ffff7c42476 libc.so.6`raise + 22
    frame #2: 0x00007ffff7c287f3 libc.so.6`abort + 211
    frame #3: 0x00007ffff7c89676 libc.so.6`___lldb_unnamed_symbol3390 + 662
    frame #4: 0x00007ffff7d3659a libc.so.6`__fortify_fail + 42
    frame #5: 0x00007ffff7d36566 libc.so.6`__stack_chk_fail + 22
    frame #6: 0x00007ffff7f56ec5 libxdp.so.1`xdp_get_ifindex_prog_id(ifindex=<unavailable>, prog_id=<unavailable>, hw_prog_id=<unavailable>, mode=<unavailable>) at libxdp.c:2572:1
    frame #7: 0x00007ffff7f5b692 libxdp.so.1`xdp_multiprog__get_from_ifindex(ifindex=1) at libxdp.c:2583:8
    frame #8: 0x00007ffff7f5c1e5 libxdp.so.1`xdp_program__attach_multi at libxdp.c:1789:11
    frame #9: 0x00007ffff7f5c10d libxdp.so.1`xdp_program__attach_multi(progs=0x00007fffffffe518, num_progs=1, ifindex=1, mode=XDP_MODE_NATIVE, flags=<unavailable>) at libxdp.c:1778:5
    frame #10: 0x00007ffff7f5c533 libxdp.so.1`xdp_program__attach(prog=<unavailable>, ifindex=<unavailable>, mode=<unavailable>, flags=<unavailable>) at libxdp.c:1890:9
    frame #11: 0x000055555555f23b server`main(argc=1, argv=0x00007fffffffe678) at server.c:44:15
    frame #12: 0x00007ffff7c29d90 libc.so.6`___lldb_unnamed_symbol3139 + 128
    frame #13: 0x00007ffff7c29e40 libc.so.6`__libc_start_main + 128
    frame #14: 0x000055555555f095 server`_start + 37
(lldb)
```

I have no idea what is going on here, is it a bug in libbpf? I've reduced the repro to be so minimal, I can't see anything that I'm doing wrong, but please let me know if this is the case.

Thanks!

**Comment by gafferongames** [comment]
I'm very sorry. It's libxdp, not libbpf.

**Comment by gafferongames** [comment]
Moving to libxdp

**Comment by gafferongames** [comment]
Hi folks, libxdp folks say it's an issue that may have been already fixed in libbpf in a patch? 

https://github.com/xdp-project/xdp-tools/issues/422

I'm using the libbpf version included in the source tree of libxdp 1.4.2, which is libbpf-1.2.0.
